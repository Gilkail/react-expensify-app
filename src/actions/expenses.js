import uuid from 'uuid'
import database from '../firebase/firebase'


// Add expense function which will trigger the addExpense action called from AddExpansePage
export const startAddExpense = (expenseData = {}) => { // Object transfered from 
    return (dispatch) => { // Transfering action dispatch due to thunk configured in store config file
        const {description = '', note = '', amount = 0, createdAt = 0} = expenseData // Destructuring object and setting up defaults
        const expense = {description, note, amount, createdAt} // Passing variables to expense array
        return database.ref('expenses').push(expense) // Pushing the data to firebase
            .then((ref)=>{ // Getting promise from fetched data with the ref
                dispatch(addExpense({ // Calling the addExpense dispatch
                    id: ref.key, // Taking the generated key from the fetched data using ref.key
                    ...expense // Spreading the expense object and passing to the action generator
                }))
        })
    }
}

// Add expnese
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

// Remove expense
export const removeExpense = ({id} = {}) =>({
    type: 'REMOVE_EXPENSE',
    id
})


// Edit expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

// export const startSetExpenses

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot)=>{
            const expenses = []
            snapshot.forEach((expense)=>{
                expenses.push({
                    id: expense.key,
                    ...expense.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}