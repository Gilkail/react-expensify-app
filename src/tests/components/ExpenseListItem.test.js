import React from 'react'
import ExpenseListItem from '../../components/ExpenseListItem'
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses'

test('Should match the expense list item', ()=>{
    const ListItemData = expenses[1]
    const wrapper = shallow(<ExpenseListItem {...ListItemData}/>)
    expect(wrapper).toMatchSnapshot()
})