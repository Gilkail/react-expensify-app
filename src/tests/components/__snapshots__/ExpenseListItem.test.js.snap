// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match the expense list item 1`] = `
<div>
  <div>
    <Link
      replace={false}
      to="/edit/2"
    >
      <h3>
        Rent
      </h3>
    </Link>
    <p>
      $0.12
      -
      28 December 1969
    </p>
  </div>
</div>
`;
